### PR TITLE
qa/cephfs: don't use sudo to write files in /tmp

### PR DESCRIPTION
**Pull request by rishabh-d-dave**
Files in /tmp cannot be written by any user( including the root user)
other than the file owner even if the permission mode on the file is
777.

Fixes: https://tracker.ceph.com/issues/49466

Marked DNM because -
Depends on https://github.com/ceph/teuthology/pull/1634.
To fix the same issue for vstart_runner.py this PR depends on https://github.com/ceph/ceph/pull/37655.

## Checklist
- [x] References tracker ticket

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by rishabh-d-dave** [comment]
Got this flake8 error (from make check) fixed locally -

```
./tasks/cephfs/cephfs_test_case.py:7:1: F401 'io.StringIO' imported but unused
```

I'll update the PR after this job finishes running - http://pulpito.front.sepia.ceph.com/rishabh-2021-03-26_11:20:12-fs-wip-rishabh-vr-set-shell-to-True-distro-basic-smithi/

File: qa/tasks/vstart_runner.py [hunk 1]
```diff
@@ -352,6 +352,23 @@ def sudo_write_file(self, path, data, **kwargs):
         """
         self.write_file(path, data, sudo=True, **kwargs)
 
+    def write_temp_file(self, **kwargs):
+        """
+        This is a convenience method that call write_file() but without sudo.
+        Reason for excluding sudo is that files in /tmp can not be written by
+        any user (including the root user) besides the file owner.
+
```
Comment (jtlayton) [inline qa/tasks/vstart_runner.py:360]: Huh. I guess this is some of the new "hardening" patches, like this one:

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=30aba6656f61ed44cba445a3c0d38b296fa9e8f5

Sigh. Not sure I'm a fan, but I guess that ship has sailed.

**Review by jtlayton — APPROVED** [review]

File: qa/tasks/vstart_runner.py (hunk 1 above)
Comment (rishabh-d-dave) [reply qa/tasks/vstart_runner.py:360]: Yes, that's the exact commit.

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
jenkins test api

**Comment by batrick** [comment]
jenkins test api

File: qa/tasks/cephfs/cephfs_test_case.py [hunk 2]
```diff
@@ -446,11 +444,7 @@ def create_client(self, client_id, moncap=None, osdcap=None, mdscap=None):
...
 
     def create_keyring_file(self, remote, keyring):
-        keyring_path = remote.run(args=['mktemp'], stdout=StringIO()).\
-            stdout.getvalue().strip()
-        sudo_write_file(remote, keyring_path, keyring)
-
-        # required when triggered using vstart_runner.py.
-        remote.run(args=['chmod', '644', keyring_path])
+        # mode must be 644 for vstart_runner.py.
+        keyring_path = remote.mktemp(data=keyring, mode='644')
```
Comment (batrick) [inline qa/tasks/cephfs/cephfs_test_case.py:448]:
````
2021-04-01T01:07:43.141 INFO:tasks.cephfs_test_runner:======================================================================
2021-04-01T01:07:43.147 INFO:tasks.cephfs_test_runner:ERROR: test_mount_all_caps_absent (tasks.cephfs.test_multifs_auth.TestClientsWithoutAuth)
2021-04-01T01:07:43.147 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2021-04-01T01:07:43.148 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2021-04-01T01:07:43.149 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_5c92bde19d0b1ad5e2a680f5925b00ec1f942c7d/qa/tasks/cephfs/test_multifs_auth.py", line 279, in test_mount_all_caps_absent
2021-04-01T01:07:43.149 INFO:tasks.cephfs_test_runner:    keyring)
2021-04-01T01:07:43.150 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_5c92bde19d0b1ad5e2a680f5925b00ec1f942c7d/qa/tasks/cephfs/cephfs_test_case.py", line 447, in create_keyring_file
2021-04-01T01:07:43.150 INFO:tasks.cephfs_test_runner:    keyring_path = remote.mktemp(data=keyring, mode='644')
2021-04-01T01:07:43.151 INFO:tasks.cephfs_test_runner:TypeError: mktemp() got an unexpected keyword argument 'mode'
````

From: /ceph/teuthology-archive/pdonnell-2021-04-01_00:45:34-fs-wip-pdonnell-testing-20210331.222326-distro-basic-smithi/6013903/teuthology.log

**Review by batrick — CHANGES_REQUESTED** [review]

**Comment by lxbsz** [comment]
Test the tmp file mentioned above on CentOS 8.1 with the ceph-client's latest kernel, please see bellow:

```
[root@ceph src]# uname -r
5.12.0-rc2+

[root@ceph src]# echo 11 > /tmp/tmp.RVXPYRVdIjG
[root@ceph src]# cat !$
cat /tmp/tmp.RVXPYRVdIjG
11
[root@ceph src]# ls -l /tmp/tmp.RVXPYRVdIjG
-rw-------. 1 lxb lxb 3 Apr  9 11:47 /tmp/tmp.RVXPYRVdIjG
[root@ceph src]# 
```

The root user can write to the tmp files not owned by it.

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2021-Apr-08

**Pull request merged**